job-server/src/spark.jobserver/auth/SJSAuthenticator.scala

```diff
@@ -0,0 +1,120 @@
+package spark.jobserver.auth
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext
+import spray.routing.directives.AuthMagnet
+import spray.routing.authentication.UserPass
+import spray.routing.authentication.BasicAuth
+
+import spray.routing.authentication._
+import spray.routing.directives.AuthMagnet
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, ExecutionContext, Future }
+import org.slf4j.LoggerFactory
+import com.typesafe.config.Config
+import org.apache.shiro.SecurityUtils
+import org.apache.shiro.authc._
+import org.apache.shiro.config.IniSecurityManagerFactory
+import org.apache.shiro.mgt.SecurityManager
+import org.apache.shiro.util.Factory
+import org.apache.shiro.subject.Subject
+
+trait SJSAuthenticator {
+
+  def asShiroAuthenticator(config: Config)(implicit ec: ExecutionContext): AuthMagnet[AuthInfo] = {
+    val logger = LoggerFactory.getLogger(getClass)
+
+    //TODO - can we read this information from the config instead?
+    val ldapFactory = new IniSecurityManagerFactory(config.getString("shiro.path.ini"))
+    val sManager = ldapFactory.getInstance()
+    SecurityUtils.setSecurityManager(sManager)
+
+    val allowedGroups: Option[java.util.List[String]] = {
+      if (config.hasPath("shiro.ldap.allowedGroups")) {
+        val g = config.getStringList("shiro.ldap.allowedGroups")
+        if (g.size() > 0) { Some(g) } else { None }
+      } else {
+        None
+      }
+    }
+
+    def isInAllowedGroupOrNoCheckOnGroups(currentUser: Subject): Boolean = {
+      allowedGroups match {
+        case Some(groups) => {
+          val roles: Array[Boolean] = currentUser.hasRoles(groups)
+          roles.foldLeft(false)((r, isInGroup) => isInGroup || r)
+        }
+        case None => {
+          true
+        }
+      }
+    }
+
+    def validate(userPass: Option[UserPass]): Future[Option[AuthInfo]] = {
+      val currentUser = SecurityUtils.getSubject()
+
+      //if (!currentUser.isAuthenticated()) {
+      Future {
+        val UserPass(user, pass) = userPass.get
+        val token = new UsernamePasswordToken(user, pass)
+        try {
+          currentUser.login(token)
+          val fullName = currentUser.getPrincipal().toString
+          if (isInAllowedGroupOrNoCheckOnGroups(currentUser)) {
+            logger.trace("ACCESS GRANTED for user [" + fullName + "]")
+            currentUser.logout()
+            Option(new AuthInfo(new User(fullName)))
+          } else {
+            currentUser.logout()
+            logger.info("ACCESS DENIED (GROUP)")
+            None
+          }
+        } catch {
+          case uae: UnknownAccountException => {
+            logger.info("Unknown user")
+            None
+          }
+          case ice: IncorrectCredentialsException => {

```

Comment:
Braces are not needed for case statements, I think it looks better without them....